File: frontend/src/services/notificationService.ts
/**
 * Notification Service
 * Handles real-time notifications via Supabase and Django backend
 */

import { supabase } from '../lib/supabase';
import api from './api';
import type { RealtimeChannel } from '@supabase/supabase-js';

export interface Notification {
  id: string;
  user_id?: string;
  username?: string;
  title: string;
  message: string;
  notification_type: 'info' | 'success' | 'warning' | 'error' | 'task' | 'user' | 'system';
  is_read: boolean;
  read_at?: string;
  related_model?: string;
  related_id?: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

type NotificationCallback = (notification: Notification) => void;

class NotificationService {
  private channel: RealtimeChannel | null = null;
  private listeners: NotificationCallback[] = [];

  /**
   * Initialize real-time notifications for a user
   */
  async initialize(userId: string): Promise<void> {

    // Unsubscribe from any existing channel
    if (this.channel) {
      await this.channel.unsubscribe();
    }

    // Create new channel for user's notifications
    this.channel = supabase
      .channel(`notifications:user_id=eq.${userId}`)
      .on(
        'postgres_changes',
        {
          event: 'INSERT',
          schema: 'public',
          table: 'notifications',
          filter: `user_id=eq.${userId}`,
        },
        (payload) => {
          console.log('🔔 New notification received:', payload);
          const notification = payload.new as Notification;
          this.notifyListeners(notification);
        }
      )
      .on(
        'postgres_changes',
        {
          event: 'UPDATE',
          schema: 'public',
          table: 'notifications',
          filter: `user_id=eq.${userId}`,
        },
        (payload) => {
          console.log('🔔 Notification updated:', payload);
          const notification = payload.new as Notification;
          this.notifyListeners(notification);
        }
      )
      .subscribe((status) => {
        console.log('📡 Notification channel status:', status);
      });
  }

  /**
   * Clean up and unsubscribe
   */
  async cleanup(): Promise<void> {
    if (this.channel) {
      await this.channel.unsubscribe();
      this.channel = null;
    }
    this.listeners = [];
  }

  /**
   * Subscribe to notification events
   */
  subscribe(callback: NotificationCallback): () => void {
    this.listeners.push(callback);

    // Return unsubscribe function
    return () => {
      this.listeners = this.listeners.filter((cb) => cb !== callback);
    };
  }

  /**
   * Notify all listeners of new notification
   */
  private notifyListeners(notification: Notification): void {
    this.listeners.forEach((callback) => {
      try {
        callback(notification);
      } catch (error) {
        console.error('Error in notification listener:', error);
      }
    });
  }

  /**
   * Fetch all notifications from Django backend
   */
  async getAll(): Promise<Notification[]> {
    try {
      const response = await api.get('/notifications/');
      // Handle paginated response from DRF
      if (response.data && Array.isArray(response.data.results)) {
        return response.data.results;
      }
      // Handle non-paginated response (plain array)
      if (Array.isArray(response.data)) {
        return response.data;
      }
      // Fallback to empty array
      console.warn('Unexpected notification response format:', response.data);
      return [];
    } catch (error) {
      console.error('Failed to fetch notifications:', error);
      throw error;
    }
  }

  /**
   * Get unread notification count
   */
  async getUnreadCount(): Promise<number> {
    try {
      const response = await api.get('/notifications/unread_count/');
      return response.data.unread_count;
    } catch (error) {
      console.error('Failed to fetch unread count:', error);
      return 0;
    }
  }

  /**
   * Mark a notification as read
   */
  async markAsRead(notificationId: string): Promise<void> {
    try {
      await api.post(`/notifications/${notificationId}/mark_read/`);
    } catch (error) {
      console.error('Failed to mark notification as read:', error);
      throw error;
    }
  }

  /**
   * Mark all notifications as read
   */
  async markAllAsRead(): Promise<void> {
    try {
      await api.post('/notifications/mark_all_read/');
    } catch (error) {
      console.error('Failed to mark all as read:', error);
      throw error;
    }
  }

  /**
   * Create a new notification (admin/system only)
   */
  async create(notification: Partial<Notification>): Promise<Notification> {
    try {
      const response = await api.post('/notifications/', notification);
      return response.data;
    } catch (error) {
      console.error('Failed to create notification:', error);
      throw error;
    }
  }

  /**
   * Delete a notification
   */
  async delete(notificationId: string): Promise<void> {
    try {
      await api.delete(`/notifications/${notificationId}/`);
    } catch (error) {
      console.error('Failed to delete notification:', error);
      throw error;
    }
  }

  /**
   * Create a notification in Supabase (for real-time delivery)
   * This bypasses Django and directly inserts into Supabase
   */
  async createRealtime(notification: Partial<Notification>): Promise<void> {
    try {
      const { data: { user } } = await supabase.auth.getUser();

      if (!user) {
        throw new Error('User not authenticated');
      }

      const { error } = await supabase
        .from('notifications')
        .insert({
          user_id: notification.user_id || user.id,
          username: notification.username,
          title: notification.title,
          message: notification.message,
          notification_type: notification.notification_type || 'info',
          is_read: false,
          related_model: notification.related_model,
          related_id: notification.related_id,
          metadata: notification.metadata || {},
        });

      if (error) throw error;
    } catch (error) {
      console.error('Failed to create realtime notification:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const notificationService = new NotificationService();
export default notificationService;

File: backend/api/services/notification_service.py
"""
Notification Service

This module handles all notification-related business logic including:
- Notification creation
- Notification delivery
- Bulk notification sending
- Notification management
"""

from typing import Dict, Any, Optional, List
from django.contrib.auth.models import User
from django.db import transaction
from django.utils import timezone
from api.models import Notification, Organization
from .base_service import BaseService, ServiceException


class NotificationService(BaseService):
    """Service class for notification-related operations"""

    @transaction.atomic
    def create_notification(
        self,
        user: User,
        title: str,
        message: str,
        notification_type: str = 'info',
        **metadata
    ) -> Notification:
        """
        Create a notification for a user

        Args:
            user: User to send notification to
            title: Notification title
            message: Notification message
            notification_type: Type of notification (info, success, warning, error, task, user, system)
            **metadata: Additional metadata

        Returns:
            Created Notification instance

        Raises:
            ServiceException: If creation fails
        """
        try:
            # Validate notification type
            valid_types = [choice[0] for choice in Notification.TYPE_CHOICES]
            if notification_type not in valid_types:
                raise ServiceException(
                    message=f"Invalid notification type: {notification_type}",
                    code='invalid_type',
                    status=400
                )

            # Create notification
            notification = Notification.objects.create(
                user=user,
                title=title,
                message=message,
                notification_type=notification_type,
                metadata=metadata
            )

            self.log_action(
                f"Created notification for user: {user.username}",
                {'title': title, 'type': notification_type}
            )

            return notification

        except ServiceException:
            raise
        except Exception as e:
            self.logger.error(f"Failed to create notification: {str(e)}")
            raise ServiceException(
                message=f"Failed to create notification: {str(e)}",
                code='notification_creation_failed',
                status=500
            )

    @transaction.atomic
    def create_bulk_notifications(
        self,
        users: List[User],
        title: str,
        message: str,
        notification_type: str = 'info',
        **metadata
    ) -> List[Notification]:
        """
        Create notifications for multiple users

        Args:
            users: List of users to send notification to
            title: Notification title
            message: Notification message
            notification_type: Type of notification
            **metadata: Additional metadata

        Returns:
            List of created Notification instances

        Raises:
            ServiceException: If creation fails
        """
        try:
            notifications = []

            for user in users:
                notification = Notification.objects.create(
                    user=user,
                    title=title,
                    message=message,
                    notification_type=notification_type,
                    metadata=metadata
                )
                notifications.append(notification)

            self.log_action(
                f"Created bulk notifications",
                {'user_count': len(users), 'title': title}
            )

            return notifications

        except Exception as e:
            self.logger.error(f"Failed to create bulk notifications: {str(e)}")
            raise ServiceException(
                message=f"Failed to create bulk notifications: {str(e)}",
                code='bulk_notification_failed',
                status=500
            )

    def mark_as_read(
        self,
        notification: Notification
    ) -> Notification:
        """
        Mark a notification as read

        Args:
            notification: Notification to mark as read

        Returns:
            Updated Notification instance

        Raises:
            ServiceException: If operation fails
        """
        try:
            notification.is_read = True
            notification.read_at = timezone.now()
            notification.save()

            self.log_action(
                f"Marked notification as read",
                {'notification_id': notification.id}
            )

            return notification

        except Exception as e:
            self.logger.error(f"Failed to mark notification as read: {str(e)}")
            raise ServiceException(
                message=f"Failed to mark as read: {str(e)}",
                code='mark_read_failed',
                status=500
            )

    @transaction.atomic
    def mark_all_as_read(
        self,
        user: User
    ) -> int:
        """
        Mark all unread notifications as read for a user

        Args:
            user: User to mark notifications for

        Returns:
            Number of notifications marked as read

        Raises:
            ServiceException: If operation fails
        """
        try:
            count = Notification.objects.filter(
                user=user,
                is_read=False
            ).update(
                is_read=True,
                read_at=timezone.now()
            )

            self.log_action(
                f"Marked all notifications as read for user: {user.username}",
                {'count': count}
            )

            return count

        except Exception as e:
            self.logger.error(f"Failed to mark all as read: {str(e)}")
            raise ServiceException(
                message=f"Failed to mark all as read: {str(e)}",
                code='mark_all_read_failed',
                status=500
            )

    def get_unread_count(
        self,
        user: User
    ) -> int:
        """
        Get count of unread notifications for a user

        Args:
            user: User to get count for

        Returns:
            Number of unread notifications

        Raises:
            ServiceException: If operation fails
        """
        try:
            count = Notification.objects.filter(
                user=user,
                is_read=False
            ).count()

            return count

        except Exception as e:
            self.logger.error(f"Failed to get unread count: {str(e)}")
            raise ServiceException(
                message=f"Failed to get unread count: {str(e)}",
                code='count_failed',
                status=500
            )
